### PR TITLE
Edit groups

### PR DESCRIPTION
**Pull request by szabinst**
**Changes**: 

- Renamed CourseList related views and subviews to AllCourses
- New Core Data entities for courses and for groups on AllCourses screen. 
- New model objects for courses and groups so we don't need to deal with NSManagedObjects. 
- New getGroup API request that returns favorite groups on the Dashboard. 

**What's not included:** 
- Select all button for courses and for groups (like on Android) 
 
## Screenshots

<table>
<tr><th>Courses</th><th>Groups</th></tr>
<tr>
<td></td>
<td></td>
</tr>
</table>

Edit groups

refs: MBL-16636
affects: Student, Teacher
release note: Added the ability to edit groups on the dashboard
test plan: See ticket

## Checklist

- [x] A11y checked
- [x] Tested on phone
- [x] Tested on tablet
- [x] Tested in dark mode
- [x] Tested in light mode
- [ ] Approve from product

### CODE REVIEW

**Comment by inst-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Please add a release note. If no release note is wanted, use `none`. Example: `release note: Fixed a bug that prevented users from enjoying the app.`
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  Please add which apps this change affects. Example: `affects: Teacher, Student` or `affects: none`
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  Please add a reference to a jira ticket. For example: `refs: MBL-10023`
  </td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>One or more files are below the minimum test coverage 50%</td>
    </tr>
  </tbody>
</table>


[MBL-16636](https://instructure.atlassian.net/browse/MBL-16636)

Coverage | New % | Master % | Delta
-------- | ----- | -------- | -----
**Canvas iOS** | 92.34% | 92.38% | -0.04%
Core/Core/Groups/GetGroups.swift | 47.73% | 64.62% | -16.89%
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against aae2f6226a1f421532e908a55830697539f9c2ed
</p>

**Comment by szabinst** [comment]
<details><summary>Teacher Build QR Code:</summary>

</details>

**Comment by inst-danger** [comment]
<details><summary>Teacher Build QR Code:</summary>

</details>

**Comment by inst-danger** [comment]
<details><summary>Student Build QR Code:</summary>

</details>

**Review by vargaat — CHANGES_REQUESTED** [review]
Good job on this! I only had a few minor remarks regarding usability:
- **Not in scope.** It looks strange that the future/past.current enrollments section headers are bigger than the Courses and Groups headers. We might revise this.
- **Fixed.** The last course cell looks taller than the others.
- **Fixed.** There's no separator after the group hint text while we have one at the courses section.
- **Fixed.** Account group cell's disclosure indicator is not aligned to the right.
- **Fixed.** Hint texts might need a trailing padding because it looks strange that they stick to the edge of the screen.
- **Fixed.** This was like just as before but then I pop back to the all courses screen the scroll goes to the top forgetting my previous position on the screen. I think this scrolling should only be done on the first load to hide the search bar.
- **Fixed.** Searching for a string that has no match displays the empty state instead of the no results state that also hides the search bar and I'm stuck in this state.

File: Core/Core/Courses/AllCourses/View/AllCoursesView.swift [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+                        emptyView(width: width, height: height)
+                    case .error:
+                        errorView(width: width, height: height)
+                    }
+                }
+            } refreshAction: { endRefreshing in
+                viewModel.refresh(completion: endRefreshing)
+            }
+        }
+        .avoidKeyboardArea()
```
[the remark this answers is not in the document]
Comment (szabinst) [reply Core/Core/Courses/AllCourses/View/AllCoursesView.swift:52]: I tested it, seems like it works great without it.

**Comment by szabinst** [comment]
I added the UI improvements and fixed everything. Please check again @vargaat

**Review by vargaat — DISMISSED** [review]
QA+1

**Review by balintbartok — DISMISSED** [review]
QA + 1
looks great!

**Review by domonkosadam — APPROVED** [review]
QA + 1

**Review by balintbartok — APPROVED** [review]

**Review by domonkosadam — APPROVED** [review]

**Pull request merged**